File: backend/src/main/java/com/protractor/backend/domain/schedule/entity/Schedule.java
package com.protractor.backend.domain.schedule.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import java.time.LocalDate;
import java.time.LocalDateTime;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

/**
 * 개인 일정(캘린더). 본인만 조회·수정·삭제할 수 있다.
 *
 * <p>
 * 디데이(dDayEnabled) 일정은 홈 대시보드의 D-Day 표시가 이 테이블을 읽어 쓴다.
 * 여러 개가 켜져 있으면 어떤 것을 보여줄지는 홈 담당(가장 가까운 날짜 등)이 정한다.
 */
@Entity
@Table(name = "schedules")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Schedule {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "schedule_id")
    private Long id;

    @Column(name = "member_id", nullable = false)
    private Long memberId;

    @Column(name = "title", nullable = false, length = 100)
    private String title;

    @Column(name = "target_date", nullable = false)
    private LocalDate targetDate;

    /** 캘린더 표시 색상. 형식(#16A34A 등)은 FE가 정하며 서버는 문자열로만 보관한다. */
    @Column(name = "color", length = 20)
    private String color;

    @Column(name = "d_day_enabled", nullable = false)
    private boolean dDayEnabled;

    @Column(name = "memo", length = 500)
    private String memo;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @Builder
    private Schedule(Long memberId, String title, LocalDate targetDate, String color,
                     boolean dDayEnabled, String memo) {
        this.memberId = memberId;
        this.title = title;
        this.targetDate = targetDate;
        this.color = color;
        this.dDayEnabled = dDayEnabled;
        this.memo = memo;
    }

    public boolean isOwnedBy(Long memberId) {
        return this.memberId.equals(memberId);
    }

    public void changeTitle(String title) {
        this.title = title;
    }

    public void changeTargetDate(LocalDate targetDate) {
        this.targetDate = targetDate;
    }

    /** null이면 색상 제거다. */
    public void changeColor(String color) {
        this.color = color;
    }

    public void changeDDayEnabled(boolean dDayEnabled) {
        this.dDayEnabled = dDayEnabled;
    }

    /** null이면 메모 제거다. */
    public void changeMemo(String memo) {
        this.memo = memo;
    }

    @PrePersist
    private void prePersist() {
        LocalDateTime now = LocalDateTime.now();
        this.createdAt = now;
        this.updatedAt = now;
    }

    @PreUpdate
    private void preUpdate() {
        this.updatedAt = LocalDateTime.now();
    }
}

File: frontend/src/types/schedule.ts
// 일정(캘린더) 관련 공용 타입 (API 명세서 v3 · 일정 도메인)
//
// backend domain/schedule 의 ScheduleResponse / ScheduleCreateRequest /
// ScheduleUpdateRequest 와 1:1 대응한다. 필드를 바꾸면 서버 DTO도 같이 바꿔야 한다.

/** 일정 하나. targetDate 는 서버가 LocalDate 로 다루므로 시간 성분이 없다. */
export interface Schedule {
  scheduleId: number;
  title: string;
  targetDate: string; // YYYY-MM-DD
  color?: string;
  /** 켜면 홈·캘린더에 D-day 배지가 붙는다 */
  dDayEnabled: boolean;
  memo?: string;
  createdAt: string;
  updatedAt?: string;
}

/** GET /schedules 응답. 어느 달의 응답인지 확인할 수 있게 연·월을 같이 준다. */
export interface ScheduleListResponse {
  year: number;
  month: number;
  schedules: Schedule[];
}

/** POST /schedules 요청 */
export interface ScheduleCreateRequest {
  title: string;
  targetDate: string;
  color?: string;
  dDayEnabled?: boolean;
  memo?: string;
}

/** PATCH /schedules/{id} 요청. 보낸 필드만 반영된다(부분 수정). */
export type ScheduleUpdateRequest = Partial<ScheduleCreateRequest>;

/**
 * 일정에 쓸 수 있는 색.
 *
 * 자유 입력 대신 목록으로 고정한다. 사용자가 배경과 구분되지 않는 색을 고르면 캘린더에서
 * 점이 보이지 않고, 그건 고르는 순간에는 알 수 없다. 서버는 20자 이하면 무엇이든 받는다.
 */
export const SCHEDULE_COLORS = [
  '#4f46e5',
  '#0ea5e9',
  '#16a34a',
  '#f59e0b',
  '#f04438',
  '#a855f7',
] as const;

export const DEFAULT_SCHEDULE_COLOR = SCHEDULE_COLORS[0];
